Add demux

rcywongaa opened the pull request:
Completes https://github.com/ros-tooling/topic_tools/issues/105

This PR porting https://github.com/ros/ros_comm/blob/noetic-devel/tools/topic_tools/src/demux.cpp to the ROS 2.

> demux is a generic ROS topic demultiplexer: one input topic is fanned out to 1 of N output topics. A service is provided to select between the outputs

Design choices:
- Callbacks use lambda instead of `std::bind` to get default arguments working

rcywongaa committed:
Copy, rename and make buildable

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Swap input and output topic

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Reset publisher if output topic changes

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Add tests

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Fix test failures

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Fix header guard

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Make linter happy

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Add README entry

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Remove lazy parameter

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Rename default input topic to ~/input

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Fix typo

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa commented:
The test fail doesn't seem related to my changes so this PR is considered done.  Let me know otherwise.  Thanks!

MichaelOrlov commented:
@rcywongaa Actually `Build_and_test` job failure relates to the usage of the `ament_auto_add_library` and `ament_auto_add_executable` in your PR. In general need to avoid those "magic" macros with `amen_auto`. At least we are not using them in the ROS 2 core packages.

rcywongaa committed:
Remove usage of ament_cmake_auto

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa commented:
@MichaelOrlov Ah ok, I've fixed that and changed all the other stuff to not use `ament_cmake_auto`.

Any reason why `ament_cmake_auto` is discouraged?  Seems quite nifty

MichaelOrlov commented:
> Any reason why ament_cmake_auto is discouraged? Seems quite nifty

It is an evil at the end and creates a lot of mess since we don't control what it includes as dependencies and what it exports as dependencies. Originally, it was introduced for quick prototyping but not for production-ready code.

MichaelOrlov commented on topic_tools_interfaces/package.xml:
```diff
@@ -8,7 +8,6 @@
   <maintainer email="ros-tooling@googlegroups.com">ROS Tooling Working Group</maintainer>
   <license>Apache License 2.0</license>
 
-  <buildtool_depend>ament_cmake_auto</buildtool_depend>
```
It is ok to remove the `<buildtool_depend>ament_cmake_auto</buildtool_depend>`
Although, it seems we need to add 
```
<buildtool_depend>ament_cmake</buildtool_depend>
```
to avoid compilation errors as
```
CMake Error at CMakeLists.txt:14 (find_package):
2024-06-08T19:39:10.7326444Z   By not providing "Findrosidl_default_generators.cmake" in CMAKE_MODULE_PATH
2024-06-08T19:39:10.7327166Z   this project has asked CMake to find a package configuration file provided
2024-06-08T19:39:10.7327792Z   by "rosidl_default_generators", but CMake did not find one.
```
Also will need to add `find_package(ament_cmake REQUIRED)` before `find_package(rosidl_default_generators REQUIRED)` in Cmake file.

rcywongaa replied:
Fixed, also added it for `topic_tools`

Strange that it builds without it on my PC...

MichaelOrlov replied:
I agree. It is even more strange that it fails with a similar error message even after adding a proper reference to the `ament_cmake.` But now it can not find the `ament_cmake` itself.
```
CMake Error at CMakeLists.txt:14 (find_package):
2024-06-10T19:43:53.5674842Z   By not providing "Findament_cmake.cmake" in CMAKE_MODULE_PATH this project
2024-06-10T19:43:53.5675606Z   has asked CMake to find a package configuration file provided by
2024-06-10T19:43:53.5676188Z   "ament_cmake", but CMake did not find one.
2024-06-10T19:43:53.5676497Z 
```
@emersonknapp @clalancette Any thoughts on this CMake failure?

rcywongaa replied:
Does `ament-cmake` need to be added to https://github.com/ros-tooling/setup-ros-docker/blob/main/setup-ros.sh?

Either that, or maybe `source /opt/ros/.../setup.bash` before the build?
I get the same error if I don't source ros.
I briefly tried looking through the CI pipeline and couldn't quite find where ros is sourced before the build.

Side question: What's the rationale behind doing some stuff in the action (e.g. [installing ROS](https://github.com/ros-tooling/setup-ros/blob/master/src/setup-ros-linux.ts#L189)) and some stuff in the container (e.g. [installing dev packages](https://github.com/ros-tooling/setup-ros-docker/blob/main/setup-ros.sh))?

christophebedard replied:
I believe this might be because CI is using Ubuntu Jammy, which doesn't contain packages for Rolling anymore: https://github.com/ros-tooling/topic_tools/actions/runs/9432005584/job/26042241001?pr=106#step:2:18. It should use Noble here:

- https://github.com/ros-tooling/topic_tools/blob/820a85bab1e7939c8b2f18617b48a2b7346ce74a/.github/workflows/test.yml#L16
- https://github.com/ros-tooling/topic_tools/blob/820a85bab1e7939c8b2f18617b48a2b7346ce74a/.github/workflows/lint.yml#L13

And `rolling` here: https://github.com/ros-tooling/topic_tools/blob/820a85bab1e7939c8b2f18617b48a2b7346ce74a/.github/workflows/test.yml#L20.

MichaelOrlov replied:
- Fixed in #107

rcywongaa committed:
Add missing dependency on ament_cmake

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

MichaelOrlov commented:
@rcywongaa Could you please address Uncrustify warnings?
```
Code style divergence in file 'src/demux_node.cpp':
2024-06-11T21:44:52.5204969Z     
2024-06-11T21:44:52.5205315Z     --- src/demux_node.cpp
2024-06-11T21:44:52.5205743Z     +++ src/demux_node.cpp.uncrustify
2024-06-11T21:44:52.5206238Z     @@ -116 +116 @@
2024-06-11T21:44:52.5206846Z     -      get_node_topics_interface()->resolve_topic_name(request->topic);
2024-06-11T21:44:52.5207776Z     +             get_node_topics_interface()->resolve_topic_name(request->topic);
2024-06-11T21:44:52.5208459Z     @@ -155 +155 @@
2024-06-11T21:44:52.5209056Z     -      get_node_topics_interface()->resolve_topic_name(output_topic_);
2024-06-11T21:44:52.5209947Z     +             get_node_topics_interface()->resolve_topic_name(output_topic_);
2024-06-11T21:44:52.5210980Z     @@ -172 +172 @@
2024-06-11T21:44:52.5211849Z     -        get_node_topics_interface()->resolve_topic_name(request->topic);
2024-06-11T21:44:52.5212766Z     +               get_node_topics_interface()->resolve_topic_name(request->topic);
2024-06-11T21:44:52.5213401Z     
```

rcywongaa committed:
Fix uncrustify warnings

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa commented:
Fixed, strangely, it doesn't complain locally or on [my fork](https://github.com/rcywongaa/topic_tools/pull/1)...

MichaelOrlov commented on topic_tools/test/test_topic_tool.hpp:
```diff
@@ -176,3 +184,64 @@ class TestTopicToolMultiSub : public TestTopicTool
+    auto message = std_msgs::msg::String();
+    message.data = msg_content;
+    publisher_->publish(message);
+    rclcpp::spin_some(target_node);
+    rclcpp::spin_some(test_node_);
```
Using `rclcpp::spin_some(node)` is sort of non-deterministic way to wait until some job will be done. The tests with such approach are tend o be flaky.
IMO it would be more robust to use `rclcpp:spin_until_future_complete(..)` or spin nodes in executor continuously in a separate thread until end of the test.

rcywongaa replied:
This was following the pattern from [other](https://github.com/rcywongaa/topic_tools/blob/add_demux/topic_tools/test/test_topic_tool.hpp#L97-L98)  [tests](https://github.com/rcywongaa/topic_tools/blob/add_demux/topic_tools/test/test_topic_tool.hpp#L158-L159).
I believe spinning nodes continuously in a separate thread would be similarly non-deterministic.

Would you mind to elaborate a bit more on the `spin_until_future_complete` way?  I don't think there's a way to get a `future` out of the `publish()` atm?

I think this boils down to the asynchronous nature of `publish`.  The discussion about a synchronous version of `publish()` [here](https://github.com/ros2/ros2/issues/255) and [eProsima FastDDS's synchronous publishing](https://discourse.ros.org/t/rmw-fast-dds-publication-mode-sync-vs-async-and-how-to-change-it/17153) might be relevant

In particular, the following section from [FastDDS](https://www.eprosima.com/index.php/resources-all/performance/dds-asynchronous-vs-synchronous-publishing)
> Additionally, if there are local readers (Intraprocess), the data is directly sent to each one (no transport is involved), their listeners are notified, and the corresponding callbacks are executed before returning control to the user.

seems like what we need for having deterministic callbacks.

MichaelOrlov replied:
Ahh never mind. There is too much to rework, and since other tests follow the same design pattern in this package, let's go with the current solution. 
Will rework in a follow up PR if CI will be unstable.

As regards 
> Would you mind to elaborate a bit more on the spin_until_future_complete way? I don't think there's a way to get a future out of the publish() atm?

I meant that the future could be an async task where we could check the number of messages that arrived during the timeout, for example.

A good example of how to orchestrate such a helper class for subscriptions can be found in the rosbag2 https://github.com/ros2/rosbag2/blob/rolling/rosbag2_test_common/include/rosbag2_test_common/subscription_manager.hpp

MichaelOrlov commented on topic_tools/src/demux_node.cpp:
```diff
@@ -0,0 +1,186 @@
+    output_topic_ = output_topics_.front();
+  }
+
+  discovery_timer_ = this->create_wall_timer(
+    discovery_period_, std::bind(&DemuxNode::make_subscribe_unsubscribe_decisions, this));
```
What value is assigned to the `discovery_period_`?
 I wasn't able to find where it is declared or assigned.

rcywongaa replied:
It's assigned in the base class [`ToolBaseNode`](https://github.com/rcywongaa/topic_tools/blob/add_demux/topic_tools/include/topic_tools/tool_base_node.hpp#L43), currently it's `std::chrono::milliseconds{100};`

MichaelOrlov commented on topic_tools/src/demux_node.cpp:
```diff
@@ -0,0 +1,186 @@
+  input_topic_ = declare_parameter("input_topic", "~/input");
+  lazy_ = false;
+  output_topics_ = declare_parameter<std::vector<std::string>>("output_topics");
+  if (output_topic_.empty()) {
+    output_topic_ = output_topics_.front();
+  }
```
Assigning `output_topic_ = output_topics_.front();` when `output_topic_.empty()` seems not a good idea, since `front()` will be pointing out to the ambient address in memory.
According to the https://en.cppreference.com/w/cpp/container/vector/front
> Calling front on an empty container causes undefined behavior.

christophebedard replied:
it's a bit confusing, but it's assigning something to `output_topic_` if it's empty (`output_topic_.empty()`). It's not using `output_topics_.front()` when `output_topics_.empty()`.

MichaelOrlov replied:
Yeah, To avoid confusion it would be better to have 3 variables `initial_topic`, `output_topics` and final `output_topic_`.
Also still need to check that `(!output_topics_.empty())` before using `output_topic_.front()`

rcywongaa replied:
Added `initial_topic` variable for both mux and demux

MichaelOrlov commented on README.md:
```diff
@@ -208,6 +209,37 @@ ros2 run topic_tools mux sel_cmdvel auto_cmdvel joystick_cmdvel
+
+#### Parameters
+
+- `input_topic` (string, default=`~/input`)
+    - the same as if provided as a command line argument
+- `output_topics` (string array)
+    - the same as if provided as a command line argument
+- `initial_topic` (str, default="")
+    - Output topic to select on startup. If `__none`, start with no output topic. If unset, default to first topic in arguments
+
```
Having all 3 arguments as positional when one of them is a space-separated list is not a good design decision. It will be more cleaner to make all of them as explicit no-positional arguments with short alternatives. e.g. `--input-topic -i`, `--output-topics -o`.

BTW. What is the `initial_topic` parameter? I don't see that it was parsed or taking in to account?

rcywongaa replied:
`initial_topic` is read here: https://github.com/rcywongaa/topic_tools/blob/add_demux/topic_tools/src/demux_node.cpp#L33

~~Are we open to adding dependency to a proper argparsing library?~~

~~I would avoid doing argument parsing myself, even with `getopt_long` since it could easily lead to more than half the code in the file just about parsing arguments.~~

I would argue that we should either use positional arguments (for simplicity), or force users to use the full blown ros params via `--ros-args -p param_name:=param_value`.  I don't think having something in between where we parse the arguments ourselves is the way to go.
At the moment, the convention for `topic_tools` seems to be positional arguments.

MichaelOrlov replied:
Ok, let's go with positional arguments for consistency with `mux` and other keywords from this package.

I agree that it will be better to utilize some third-party library for parsing non-positional arguments in C++. But it is much more effort to bring it in.

MichaelOrlov commented on topic_tools/src/demux.cpp:
```diff
@@ -0,0 +1,38 @@
+  auto args = rclcpp::init_and_remove_ros_arguments(argc, argv);
+  auto options = rclcpp::NodeOptions{};
+
+  if (args.size() >= 3) {
+    options.append_parameter_override("input_topic", args.at(1));
+    options.append_parameter_override(
+      "output_topics", std::vector<std::string>{args.begin() + 2, args.end()});
+  }
```
What if `args.size() == 2` ?

rcywongaa replied:
Currently,  `args.size() <= 2`, it crashes with
```
terminate called after throwing an instance of 'rclcpp::exceptions::UninitializedStaticallyTypedParameterException'
  what():  Statically typed parameter 'output_topics' must be initialized.
[ros2run]: Aborted
```

Added a helpful error message and return -1 instead

MichaelOrlov commented on README.md:
````diff
@@ -208,6 +209,37 @@ ros2 run topic_tools mux sel_cmdvel auto_cmdvel joystick_cmdvel
+
+#### Usage
+
+```shell
+ros2 run topic_tools demux <intopic> <outtopic1> [outopic2...]
+```
+
+Subscribe to <intopic1> and publish currently to selected outopic among <outopic1>...N. demux will start with <outopic1> selected.
+- `intopic`: Incoming topic to subscribe to
+- `outopicN`: Outgoing topic to publish on
````
We need to be consistent about naming. It is better to use `outtopic` with two `tt` or `out_topic` with an underscore.

rcywongaa replied:
Fixed for mux and demux

MichaelOrlov commented on topic_tools/src/tool_base_node.cpp:
```diff
@@ -30,6 +30,14 @@ ToolBaseNode::ToolBaseNode(const std::string & node_name, const rclcpp::NodeOpti
 {
   if (auto source_info = try_discover_source()) {
+    // publisher exists already but needs changing if output_topic_ changes
+    if (
+      pub_ &&
+      pub_->get_topic_name() != get_node_topics_interface()->resolve_topic_name(output_topic_))
+    {
+      pub_.reset();
+    }
```
Nitpick style suggestion.
```suggestion
    if (pub_ &&
      pub_->get_topic_name() != get_node_topics_interface()->resolve_topic_name(output_topic_))
    {
      pub_.reset();
    }
```

rcywongaa replied:
Accepted

rcywongaa committed:
Add initial_topic_ variable to avoid confusion

This addresses comment
https://github.com/ros-tooling/topic_tools/pull/106#discussion_r1637020280

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Fix outtopic spelling

This addresses comment
https://github.com/ros-tooling/topic_tools/pull/106#discussion_r1637040958

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Add error message on incorrect number of arguments

Addresses
https://github.com/ros-tooling/topic_tools/pull/106#discussion_r1637033132

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

rcywongaa committed:
Formatting

Signed-off-by: Rufus Wong <rcywongaa@gmail.com>

MichaelOrlov commented:
@rcywongaa I didn't forget about your PR. Just have a busy week, and we need to clean up the mess on a branches for `topic_tools` since we incorrectly released from rolling to the Jazzy.

Sorry for the delay, and thank you for your patience.

rcywongaa commented:
@MichaelOrlov No worries, sorry didn't mean to rush you.  Just not too sure what the standard procedure is here.  I'll avoid re-requesting a review in the future. :bow:

MichaelOrlov reviewed (commented):
@rcywongaa Thanks for the iterations and your patience.
Overall looks good now.
However, I would like you to ask to drop your [Update CI to use noble rolling](https://github.com/ros-tooling/topic_tools/pull/106/commits/f0cb259816bce2821363e68d759103e07da3e404)
commit from this PR, since we already have these changes on master.
See merged PR https://github.com/ros-tooling/topic_tools/pull/107 for details.

rcywongaa commented:
@MichaelOrlov Done!

MichaelOrlov reviewed (approved):
LGTM.

MichaelOrlov commented:
https://github.com/Mergifyio backport jazzy
